pkg/sql/drop_type.go

```diff
@@ -12,27 +12,168 @@ package sql
+	node := &dropTypeNode{
+		n:  n,
+		td: make(map[sqlbase.ID]*sqlbase.MutableTypeDescriptor),
+	}
+	if n.DropBehavior == tree.DropCascade {
+		return nil, unimplemented.NewWithIssue(51480, "DROP TYPE CASCADE is not yet supported")
+	}
+	for _, name := range n.Names {
+		// Resolve the desired type descriptor.
+		typeDesc, err := p.ResolveMutableTypeDescriptor(ctx, name, !n.IfExists)
+		if err != nil {
+			return nil, err
+		}
+		if typeDesc == nil {
+			continue
+		}
+		// If we've already seen this type, then skip it.
+		if _, ok := node.td[typeDesc.ID]; ok {
+			continue
+		}
+		// The implicit array types are not directly droppable.
+		if typeDesc.Kind == sqlbase.TypeDescriptor_ALIAS {
+			return nil, pgerror.Newf(
+				pgcode.DependentObjectsStillExist,
+				"%q is an implicit array type and cannot be modified",
+				name,
+			)
+		}
+
+		// Check if we can drop the type.
+		if err := p.canDropTypeDesc(ctx, typeDesc, n.DropBehavior); err != nil {
+			return nil, err
+		}
+
+		// Get the array type that needs to be dropped as well.
+		// TODO (rohany): This should be a method on the desc.Collection.
+		arrayDesc, err := sqlbase.GetTypeDescFromID(ctx, p.txn, p.ExecCfg().Codec, typeDesc.ArrayTypeID)
+		if err != nil {
+			return nil, err
+		}
+		mutArrayDesc := sqlbase.NewMutableExistingTypeDescriptor(*arrayDesc.TypeDesc())
+		// Ensure that we can drop the array type as well.
+		if err := p.canDropTypeDesc(ctx, mutArrayDesc, n.DropBehavior); err != nil {
+			return nil, err
+		}
+
+		// Record these descriptors for deletion.
+		node.td[typeDesc.ID] = typeDesc
+		node.td[mutArrayDesc.ID] = mutArrayDesc
+	}
+	return node, nil
+}
+
+func (p *planner) canDropTypeDesc(
+	ctx context.Context, desc *sqlbase.MutableTypeDescriptor, behavior tree.DropBehavior,
+) error {
+	// TODO (rohany): Add privilege checks here when we have them.
+	if len(desc.ReferencingDescriptorIDs) > 0 && behavior != tree.DropCascade {
+		var dependentNames []string
+		for _, id := range desc.ReferencingDescriptorIDs {
+			desc, err := p.Tables().GetMutableTableVersionByID(ctx, id, p.txn)
+			if err != nil {
+				return errors.Wrapf(err, "type has dependent objects")
+			}
+			fqName, err := p.getQualifiedTableName(ctx, desc.TableDesc())
+			if err != nil {
+				return errors.Wrapf(err, "type %q has dependent objects", desc.Name)
+			}
+			dependentNames = append(dependentNames, fqName.FQString())
+		}
+		return pgerror.Newf(
+			pgcode.DependentObjectsStillExist,
+			"cannot drop type %q because other objects (%v) still depend on it",
+			desc.Name,
+			dependentNames,
+		)
+	}
+	return nil
 }
 
 func (n *dropTypeNode) startExec(params runParams) error {
-	return errors.AssertionFailedf(
-		"should not be calling startExec on DROP TYPE node",
-	)
+	for _, typ := range n.td {
+		if err := params.p.dropTypeImpl(params.ctx, typ, tree.AsStringWithFQNames(n.n, params.Ann())); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *planner) addTypeBackReference(
```

Comment:
I think a backref also has to be added for views.

```
root@127.0.0.1:65260/movr> create type test as enum ('hello');
CREATE TYPE

Time: 2.043ms

root@127.0.0.1:65260/movr> create view vx as SELECT 'hello'::test;
CREATE VIEW

Time: 3.804ms

root@127.0.0.1:65260/movr> drop type test;
DROP TYPE

Time: 27.613ms

root@127.0.0.1:65260/movr> create view vx as SELECT 'hello'::test;
root@127.0.0.1:65260/movr> select * from vx;
ERROR: type ID 59 in descriptor not found: descriptor not found
```

Reply:
Good catch, updated.